optionally use a filter based on its existence

I propose a feature, where within a template, you can test for the existence of a filter so that you can then use that filter if it exists, and not if it doesn't.

For example,
```jinja
{% if filter is callable %}
  {{ var|filter }}
{% else %}
  {{ var }} {# no use of 'filter' #}
{% endif %}
```

Currently, if the filter does not exist, Jinja would throw a `jinja2.exceptions.TemplateAssertionError: no filter named 'filter'` because of the line `{{ var|filter }}`. The desired output is what is rendered in the `else`. You can test if the filter is callable, but you cannot then optionally use it. At a high level this is similar to how one might commonly use a var, if it exists.

```jinja
{% if myvar is defined %}
  {{ myvar }}
{% else %}
  No myvar {# no use of 'myvar' #}
{% endif %}
```

**Example Use Case:**

I would like to be able to write Lektor themes such that they can take advantage of filters provided by various Lektor plugins if they are installed. I do not want to require those plugins, and thus also their provided filters. I'd want a Lektor theme to flexibly handle either situation, as per the users' desire.

More generally, Jinja templates in any Flask or Django project could be written more flexibly to handle the optional presence filters. Adding custom or 3rd party filters is a common thing to do. Passing templates around between projects would be easier if they didn't force implementation of custom filters.

**Implementation:**

The desired effect could be achieved more than one way. For example, at least

1. not raising an exception when evaluating `{{ var|filter }}` and instead returning (basically anything) would allow the above snippet to be rendered. 

2. change the way the entire `if` is evaluated, such that the expressions contained in the contents of the "true" section are evaluated only if the test passes. Evaluation would resume upon reaching an `else` or `endif`.

My intuition was that the second option was already happening, but it is not. Though the contents of the passing section of the `if` never get to the final rendered product, the expressions are still evaluated anyway. Otherwise no exception would be rendered. If 2. happened, that could mean much nonsense could be included in between the `if` and the `else`, and not trip exceptions. That sounds like a change that's worth some careful thought, but I can't immediately think of a problem with it. I am not intimately familiar with Jinja's internals. Exceptions would still be raised if the test ever actually passes. 
Skipping that section may also mean performance improvements, which is a nice side effect.

Regardless of whether it's considered a feature or bug I guess, in 2.10 and master I also get TemplateSyntaxErrors easily. If I put many things that shouldn't work inside `{% if false%}{%  endif %}`, it breaks, which is conceptually similar to just using a non-existent filter. Jinja is trying to interpret everything, in order to find the next `else` or `endif` I think, instead of literally matching against those special strings and skipping whatever is in between.

Examples of things that fail right now:
 
- `{% if false %}{% if foo bar %}{% endif %}`; TemplateSyntaxError: expected token 'end of statement block', got 'bar'
- `{% if false %}{% foo %}{% endif %}`; TemplateSyntaxError: Encountered unknown tag 'foo'.
- `{% if false %}{{ 'foo'|bar }}{% endif %}`; TemplateAssertionError: no filter named 'bar'

An argument against changing this behavior is that the errors may be useful for debugging during normal template development, i.e., not when taking advantage of ignoring things inside an if block.

Personally, unless there's a good reason I haven't thought of yet, I'm in favor of changing the behavior. I'd say skip the execution of the if blocks when the test fails.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -39,6 +39,8 @@ Unreleased
 -   Update the template globals when calling
     ``Environment.get_template(globals=...)`` even if the template was
     already loaded. :issue:`295`
+-   Do not raise an error for undefined filters in unexecuted
+    if-statements and conditional expressions. :issue:`842`
 
 
 Version 2.11.3

--- a/src/jinja2/compiler.py
+++ b/src/jinja2/compiler.py
@@ -136,6 +136,11 @@ class Frame:
         self.loop_frame = False
         self.block_frame = False
 
+        # track whether the frame is being used in an if-statement or conditional
+        # expression as it determines which errors should be raised during runtime
+        # or compile time.
+        self.soft_frame = False
+
     def copy(self):
         """Create a copy of the current one."""
         rv = object.__new__(self.__class__)
@@ -154,10 +159,12 @@ class Frame:
         standalone thing as it shares the resources with the frame it
         was created of, but it's not a rootlevel frame any longer.
 
-        This is only used to implement if-statements.
+        This is only used to implement if-statements and conditional
+        expressions.
         """
         rv = self.copy()
         rv.rootlevel = False
+        rv.soft_frame = True
         return rv
 
     __copy__ = copy
@@ -442,7 +449,22 @@ class CodeGenerator(NodeVisitor):
             for name in getattr(visitor, dependency):
                 if name not in mapping:
                     mapping[name] = self.temporary_identifier()
+                # add check during runtime that dependencies used inside of executed
+                # blocks are defined, as this step may be skipped during compile time
+                self.writeline("try:")
+                self.indent()
                 self.writeline(f"{mapping[name]} = environment.{dependency}[{name!r}]")
+                self.outdent()
+                self.writeline("except KeyError:")
+                self.indent()
+                self.writeline("@internalcode")
+                self.writeline(f"def {mapping[name]}(*unused):")
+                self.indent()
+                self.writeline(
+                    f'raise TemplateRuntimeError("no filter named {name!r} found")'
+                )
+                self.outdent()
+                self.outdent()
 
     def enter_frame(self, frame):
         undefs = []
@@ -1641,7 +1663,7 @@ class CodeGenerator(NodeVisitor):
             self.write("await auto_await(")
         self.write(self.filters[node.name] + "(")
         func = self.environment.filters.get(node.name)
-        if func is None:
+        if func is None and not frame.soft_frame:
             self.fail(f"no filter named {node.name!r}", node.lineno)
         if getattr(func, "contextfilter", False) is True:
             self.write("context, ")
@@ -1679,6 +1701,8 @@ class CodeGenerator(NodeVisitor):
 
     @optimizeconst
     def visit_CondExpr(self, node, frame):
+        frame = frame.soft()
+
         def write_expr2():
             if node.expr2 is not None:
                 return self.visit(node.expr2, frame)

--- a/src/jinja2/runtime.py
+++ b/src/jinja2/runtime.py
@@ -36,6 +36,7 @@ exported = [
     "TemplateNotFound",
     "Namespace",
     "Undefined",
+    "internalcode",
 ]
 
 

--- a/tests/test_filters.py
+++ b/tests/test_filters.py
@@ -6,7 +6,9 @@ import pytest
 from jinja2 import Environment
 from jinja2 import Markup
 from jinja2 import StrictUndefined
+from jinja2 import TemplateRuntimeError
 from jinja2 import UndefinedError
+from jinja2.exceptions import TemplateAssertionError
 
 
 class Magic:
@@ -774,3 +776,49 @@ class TestFilter:
         t = env.from_string("{{ s|wordwrap(20) }}")
         result = t.render(s="Hello!\nThis is Jinja saying something.")
         assert result == "Hello!\nThis is Jinja saying\nsomething."
+
+    def test_filter_undefined(self, env):
+        with pytest.raises(TemplateAssertionError, match="no filter named 'f'"):
+            env.from_string("{{ var|f }}")
+
+    def test_filter_undefined_in_if(self, env):
+        t = env.from_string("{%- if x is defined -%}{{ x|f }}{%- else -%}x{% endif %}")
+        assert t.render() == "x"
+        with pytest.raises(TemplateRuntimeError, match="no filter named 'f'"):
+            t.render(x=42)
+
+    def test_filter_undefined_in_elif(self, env):
+        t = env.from_string(
+            "{%- if x is defined -%}{{ x }}{%- elif y is defined -%}"
+            "{{ y|f }}{%- else -%}foo{%- endif -%}"
+        )
+        assert t.render() == "foo"
+        with pytest.raises(TemplateRuntimeError, match="no filter named 'f'"):
+            t.render(y=42)
+
+    def test_filter_undefined_in_else(self, env):
+        t = env.from_string(
+            "{%- if x is not defined -%}foo{%- else -%}{{ x|f }}{%- endif -%}"
+        )
+        assert t.render() == "foo"
+        with pytest.raises(TemplateRuntimeError, match="no filter named 'f'"):
+            t.render(x=42)
+
+    def test_filter_undefined_in_nested_if(self, env):
+        t = env.from_string(
+            "{%- if x is not defined -%}foo{%- else -%}{%- if y "
+            "is defined -%}{{ y|f }}{%- endif -%}{{ x }}{%- endif -%}"
+        )
+        assert t.render() == "foo"
+        assert t.render(x=42) == "42"
+        with pytest.raises(TemplateRuntimeError, match="no filter named 'f'"):
+            t.render(x=24, y=42)
+
+    def test_filter_undefined_in_condexpr(self, env):
+        t1 = env.from_string("{{ x|f if x is defined else 'foo' }}")
+        t2 = env.from_string("{{ 'foo' if x is not defined else x|f }}")
+        assert t1.render() == t2.render() == "foo"
+
+        with pytest.raises(TemplateRuntimeError, match="no filter named 'f'"):
+            t1.render(x=42)
+            t2.render(x=42)

